tensorflow/lite/experimental/shlo/ops/dot_general.cc

```diff
@@ -0,0 +1,441 @@
+/* Copyright 2024 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/lite/experimental/shlo/ops/dot_general.h"
+
+#include "absl/status/status.h"
+#include "tensorflow/lite/experimental/shlo/data_type.h"
+#include "tensorflow/lite/experimental/shlo/dispatch.h"
+#include "tensorflow/lite/experimental/shlo/quantize.h"
+#include "tensorflow/lite/experimental/shlo/quantized_tensor_element_type.h"
+#include "tensorflow/lite/experimental/shlo/shape.h"
+#include "tensorflow/lite/experimental/shlo/tensor.h"
+
+namespace shlo_ref {
+
+absl::Status CheckParameters(const Tensor& lhs, const Tensor& rhs,
+                             const Tensor& lhs_batching_dimensions,
+                             const Tensor& rhs_batching_dimensions,
+                             const Tensor& lhs_contracting_dimensions,
+                             const Tensor& rhs_contracting_dimensions,
+                             Tensor& output) {
+  const int32_t* lhsb = lhs_batching_dimensions.GetDataAs<DataType::kSI32>();
+  const int32_t* rhsb = rhs_batching_dimensions.GetDataAs<DataType::kSI32>();
+  const int32_t* lhsc = lhs_contracting_dimensions.GetDataAs<DataType::kSI32>();
+  const int32_t* rhsc = rhs_contracting_dimensions.GetDataAs<DataType::kSI32>();
+  const DimensionSize lhsb_size = lhs_batching_dimensions.NumElements();
+  const DimensionSize rhsb_size = rhs_batching_dimensions.NumElements();
+  const DimensionSize lhsc_size = lhs_contracting_dimensions.NumElements();
+  const DimensionSize rhsc_size = rhs_contracting_dimensions.NumElements();
+  const size_t lhs_rank = lhs.Rank();
+  const size_t rhs_rank = rhs.Rank();
+  const size_t output_rank = output.Rank();
+  std::vector<size_t> lhs_result_dims;
```

Comment:
We try to avoid memory allocations. You can use `absl::InlinedVector<T, *REASONABLE_SIZE*>` to have a vector that uses small buffer optimisation.

Reply:
used absl::InlinedVector vectors.